walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,174 @@
+  private List<List<String>> optionalAttendeesCombs;
+
+  /** 
+   * Finds a list of times when the requested event can happen, 
+   * given the list of current events and the request information. 
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> attendees = new ArrayList<>(request.getAttendees());
+    Collection<String> optionalAttendees = new ArrayList<>(request.getOptionalAttendees());
+    // On the webapp, if form is left empty, an empty string is obtained. Remove them here.
+    attendees.remove("");
+    optionalAttendees.remove("");
+
+    long duration = request.getDuration();
+
+    // Populate {@code attendeesEventTimes} and {@code optionalAttendeesEventTimes}.
+    getAttendeesEventTimes(events, attendees, optionalAttendees);
+
+    Collection<TimeRange> possibleTimes = new ArrayList<>();
+    if (attendees.isEmpty()) {
+      // If there is no mandatory attendee, only consider availabilities of optional attendees.
+
+      // {@code eventTimes} is the list of current events' time ranges that need to be avoided when scheduling.
+      List<TimeRange> eventTimes = new ArrayList<>();
+      for (List<TimeRange> timeList : optionalAttendeesEventTimes.values()) {
+        eventTimes.addAll(timeList);
+      }
+      Collections.sort(eventTimes, TimeRange.ORDER_BY_START);
+      possibleTimes = findTimeRangeGaps(eventTimes, duration);
+    } else {
+      // If there are mandatory attendees, consider optional attendees who can possibly attend and ignore those who can't. 
+
+      List<TimeRange> eventTimes = new ArrayList<>(attendeesEventTimes);
+
+      // Find the time slot(s) that maximize the number of optional attendees who can attend.
+      // Count down from the maximum number of optional attendees and find all combinations of that number of attendees.
+      // Once we find a combination where scheduling is possible, we've found the maximal group of attendees.
+      for (int num = optionalAttendees.size(); num > 0; num--) {
+        String[] arr = optionalAttendees.toArray(new String[optionalAttendees.size()]);
+        getAllCombinations(arr, arr.length, num);
+
+        for (List<String> optionalAttendeesChosen : optionalAttendeesCombs) {
+          for (String attendee : optionalAttendeesChosen) {
+            eventTimes.addAll(optionalAttendeesEventTimes.get(attendee));
+          }
+          Collections.sort(eventTimes, TimeRange.ORDER_BY_START);
+          possibleTimes = findTimeRangeGaps(eventTimes, duration);
+          if (!possibleTimes.isEmpty()) {
+            // We've found a combination where scheduling is possible, so those are the desired time slot(s).
+            return possibleTimes;
+          }
+          // Re-initialize unavailable time ranges with those of the mandatory attendees.
+          eventTimes = new ArrayList<>(attendeesEventTimes);
```

Comment:
you can simply this complicated code by moving the declaration of eventTimes into the optionalAttendeesCombs for loop.

for (int num = optionalAttendees.size(); num > 0; num--) {
   .......
   for (List<String> optionalAttendeesChosen : optionalAttendeesCombs) {
      List<TimeRange> eventTimes = new ArrayList<>(attendeesEventTimes);
      for (String attendee : optionalAttendeesChosen) {
            eventTimes.addAll(optionalAttendeesEventTimes.get(attendee));
       }

}

so that you can remove line 63 and line 83 by combing them into one line.

Reply:
Thanks for the suggestion, but I would need `eventTimes` to be defined on line 63 and used on line 87 if no optional attendee is selected (i.e. the `for` loop from line 68 to 85 won't be executed at all).

Also, I've always been curious why variable definition in a loop doesn't count as re-definition in Java 0.0